File: iosApp/iosAppTests/Support/PublicAPI/Conversations.swift
//
//  Conversations.swift
//  MessengerUITests
//
//  Created by Morehouse, Matthew on 6/3/21.
//  Copyright © 2021 Genesys. All rights reserved.
//

import XCTest

extension ApiHelper {

    public func answerNewConversation(timeout: Int = 60) -> ConversationInfo? {
        let agentId = TestConfig.shared.config?.agentId ?? ""
        changePresence(presenceName: "On Queue", userID: agentId)
        guard let conversation = waitForConversation(timeout: timeout) else {
            print("Did not receive a new conversation to answer.")
            return nil
        }
        sendConnectOrDisconnect(conversationInfo: conversation, connecting: true, wrapup: false)
        changePresence(presenceName: "Available", userID: agentId)
        return getConversation(conversationId: conversation.conversationId)
    }

    public func disconnectExistingConversations() {
        let agentId = TestConfig.shared.config?.agentId ?? ""
        changePresence(presenceName: "Available", userID: agentId)
        guard let conversationList = getSmsConversations()?.value(forKey: "entities") as? [JsonDictionary] else {
            XCTFail("Failed to get the existing conversation list.")
            return
        }
        for conversation in conversationList {
            sendConnectOrDisconnect(conversationInfo: ConversationInfo(json: conversation), connecting: false, wrapup: true)
        }

        // Answer and disconnect any potential conversations in the queue but not yet routed to the agent.
        if let conversationInfo = answerNewConversation(timeout: 2) {
            sendConnectOrDisconnect(conversationInfo: conversationInfo, connecting: false, wrapup: true)
        }
    }

    private func getSmsConversations() -> JsonDictionary? {
        return publicAPICall(httpMethod: "GET", httpURL: "/api/v2/conversations")
    }

    private func getConversation(conversationId: String) -> ConversationInfo? {
        if let conversationJson = publicAPICall(httpMethod: "GET", httpURL: "/api/v2/conversations/\(conversationId)") {
            return ConversationInfo(json: conversationJson)
        }
        return nil
    }

    public func waitForConversation(timeout: Int = 60) -> ConversationInfo? {
        var entities: [JsonDictionary] = []
        // 1 minute wait maximum.
        // Waiting for the first conversation that matches this criteria:
        //   - An agent is on the conversation.
        //   - That agent is alerting.
        Wait.queryUntilTrue(queryInfo: "Waiting for a conversation to be available.", attempts: timeout, canFailTest: false) {
            entities = (getSmsConversations()?.value(forKey: "entities") as? [JsonDictionary] ?? []).filter {
                let agentParticipant = ($0.value(forKey: "participants") as? [JsonDictionary] ?? []).filter {
                    return $0.value(forKey: "purpose") as? String == "agent"
                }.first
                return (agentParticipant?.value(forKey: "messages") as? [JsonDictionary])?.first?.value(forKey: "state") as? String == "alerting"
            }
            return !entities.isEmpty
        }
        let conversations = entities.map {
            return ConversationInfo(json: $0)
        }.filter {
            return $0.isAvailable
        }
        return conversations.first
    }

    public func sendOutboundSmsMessage(conversationId: String, communicationId: String, message: String = "Test from agent.") {
        let payload: [String: Any] = ["textBody": message]
        _ = publicAPICall(httpMethod: "POST", httpURL: "/api/v2/conversations/messages/\(conversationId)/communications/\(communicationId)/messages", jsonBody: payload)
    }

    public func sendTypingIndicator(conversationId: String, communicationId: String) {
        let conversationEventTyping: [String: Any] = ["type": "On"]
        let payload: [String: Any] = ["typing": conversationEventTyping]
        _ = publicAPICall(httpMethod: "POST", httpURL: "/api/v2/conversations/messages/\(conversationId)/communications/\(communicationId)/typing", jsonBody: payload)
    }

    public func getCommunicationId(conversationId: String) -> String? {
        guard let results = publicAPICall(httpMethod: "GET", httpURL: "/api/v2/conversations/\(conversationId)"), let participants = results.value(forKey: "participants") as? [JsonDictionary] else {
            print("There was an issue getting the conversation information.")
            return nil
        }
        for participant in participants where participant.value(forKey: "purpose") as? String == "agent" {
            let messages = participant.value(forKey: "messages") as? [JsonDictionary] ?? []
            if let communicationId = messages[0].value(forKey: "id") as? String {
                return communicationId
            } else {
                print(participant)
            }
        }
        return nil
    }

    /// Sends the appropriate public API request to answer/disconnect from an ACD conversation.
    public func sendConnectOrDisconnect(conversationInfo: ConversationInfo, connecting: Bool, wrapup: Bool = true) {
        print("Sending request to \(connecting ? "Connect to" : "Disconnect from") a conversation.")
        let statusPayload: [String: String] = connecting ? ["state": "CONNECTED"] : ["state": "DISCONNECTED"]
        let wrapupCodeId = getDefaultWrapupCode(conversationId: conversationInfo.conversationId, participantId: conversationInfo.agentParticipantId)
        let wrapupCodePayload: [String: Any] = [
            "code": wrapupCodeId, /// This is the ID of an existing wrapup code in the Supervisor org.
            "notes": ""
        ]

        /// Send requests to disconnect the conversation and send the wrapup code.
        _ = sendStatusToParticipant(conversationId: conversationInfo.conversationId, participantId: conversationInfo.agentParticipantId, statusPayload: statusPayload)
        if connecting {
            waitForParticipantToConnect(conversationId: conversationInfo.conversationId)
        } else if wrapup {
            Wait.delay(2)
            print("Sending wrapup code.")
            sendWrapupCode(conversationInfo: conversationInfo, wrapupPayload: wrapupCodePayload)
        }
    }

    public func sendStatusToParticipant(conversationId: String, participantId: String, statusPayload: [String: Any]) -> [String: Any]? {
        return publicAPICall(httpMethod: "PATCH", httpURL: "/api/v2/conversations/messages/\(conversationId)/participants/\(participantId)", jsonBody: statusPayload)
    }

    private func waitForParticipantToConnect(conversationId: String) {
        Wait.queryUntilTrue {
            guard let conversationInfo = getConversation(conversationId: conversationId) else {
                return false
            }
            return (conversationInfo.agentParticipant?.value(forKey: "messages") as? [JsonDictionary])?.first?.value(forKey: "state") as? String == "connected"
        }
    }

    public func getDefaultWrapupCode(conversationId: String, participantId: String) -> String {
        return publicAPICall_Array(httpMethod: "GET", httpURL: "/api/v2/conversations/messages/\(conversationId)/participants/\(participantId)/wrapupcodes")?.first?.value(forKey: "id") as? String ?? ""
    }

    public func sendWrapupCode(conversationInfo: ConversationInfo, wrapupPayload: JsonDictionary) {
        _ = publicAPICall(httpMethod: "POST", httpURL: "/api/v2/conversations/messages/\(conversationInfo.conversationId)/participants/\(conversationInfo.agentParticipantId)/communications/\(conversationInfo.communicationId)/wrapup", jsonBody: wrapupPayload)
    }

    public func sendOutboundSmsImage(conversationId: String, communicationId: String) {
        guard let mediaId = uploadFileForMessages(conversationId: conversationId, communicationId: communicationId) else {
            print("Failed to upload the image.")
            return
        }
        let json: [String: Any] = [
            "direction": "outbound",
            "mediaIds": [mediaId]
        ]
        _ = publicAPICall(httpMethod: "POST", httpURL: "/api/v2/conversations/messages/\(conversationId)/communications/\(communicationId)/messages", jsonBody: json)
    }

    private func uploadFileForMessages(conversationId: String, communicationId: String) -> String? {
        // Prepare backend for a new file. Get the Upload URL.
        let result = publicAPICall(httpMethod: "POST", httpURL: "/api/v2/conversations/messages/\(conversationId)/communications/\(communicationId)/messages/media")
        guard let mediaId = result?.value(forKey: "id") as? String else {
            print("Failed to get the upload Url for messages.")
            return nil
        }

        // Send the file.
        guard let imageData = UIImage(named: "image")?.pngData() else {
            XCTFail("Failed to get Image information.")
            return nil
        }
        _ = publicAPICall(httpMethod: "POST", url: "https://apps.inindca.com/uploads/v4/services/messaging/\(mediaId)", imageData: imageData)

        // Wait for the file upload to finish.
        for _ in 0...10 {
            let check = publicAPICall(httpMethod: "GET", httpURL: "/api/v2/conversations/messages/\(conversationId)/communications/\(communicationId)/messages/media/\(mediaId)")
            print(check ?? "N/A")
            if check?.value(forKey: "status") as? String == "valid" {
                break
            }
            sleep(1) // Small delay.
        }

        return mediaId
    }
}

File: androidComposePrototype/src/androidTest/java/com/genesys/cloud/messenger/uitest/support/ApiHelper/Conversations.kt
package com.genesys.cloud.messenger.uitest.support.ApiHelper

import android.util.Log
import com.genesys.cloud.messenger.androidcomposeprototype.ui.testbed.TestBedViewModel
import com.genesys.cloud.messenger.uitest.support.testConfig
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.lang.Thread.sleep

private val TAG = TestBedViewModel::class.simpleName

@Serializable
data class Conversation(
    val id: String,
    var participants: Array<Participant>
) {
    fun getParticipantForUserId(userId: String): Participant? {
        return participants.firstOrNull { participant ->
            participant.userId == userId &&
                participant.purpose == "agent" &&
                participant.messages
                    .firstOrNull()
                    ?.isDisconnected()
                    ?.not() ?: false
        }
    }

    fun getParticipantFromPurpose(purpose: String): Participant? {
        return participants.firstOrNull { participant ->
            participant.purpose == purpose &&
                participant.messages
                    .firstOrNull()
                    ?.isDisconnected()
                    ?.not() ?: false
        }
    }

    fun getCommunicationId(participant: Participant): String {
        return participant.messages[0].id
    }
}

@Serializable
data class Participant(
    val id: String,
    val purpose: String,
    val userId: String?,
    var messages: Array<CallDetails>
)

@Serializable
data class Media(
    val id: String,
    val uploadUrl: String,
    val Status: String
)

@Serializable
data class CallDetails(
    val state: String,
    val id: String
) {
    fun isConnected(): Boolean = state == "connected"

    fun isDisconnected(): Boolean = state == "disconnected" || state == "terminated"

    fun isAlerting(): Boolean = state == "alerting"
}

fun API.waitForConversation(): Conversation? {
    for (x in 0..60) {
        val conversations = getAllConversations()
        if (conversations != null) {
            conversations.forEach { conversation ->
                Log.i(TAG, "conversationId: $conversation.id")
                val callDetails = conversation.getParticipantFromPurpose("agent")?.messages
                if (callDetails != null) {
                    callDetails.forEach { callDetail ->
                        Log.i(TAG, "call detail state: $callDetail.state")
                        if (callDetail.isAlerting()) {
                            return conversation
                        }
                    }
                }
            }
        }
        sleep(1000)
    }
    return null
}

fun API.getAllConversations(): Array<Conversation> {
    val entities = publicApiCall("GET", "/api/v2/conversations")
        ?.jsonObject?.get("entities")?.jsonArray
    return parseJsonToClass(entities?.toList())
}

fun API.getConversationInfo(conversationId: String): Conversation {
    val result = publicApiCall("GET", "/api/v2/conversations/$conversationId")
    return parseJsonToClass(result)
}

fun API.sendTypingIndicatorFromAgentToUser(conversationInfo: Conversation) {
    val agentParticipant = conversationInfo.getParticipantFromPurpose("agent")
    val participant = conversationInfo.getParticipantForUserId(agentParticipant!!.userId!!)
    val payload = "{\"typing\": {\"type\": \"On\"}}".toByteArray()
    publicApiCall("POST", "/api/v2/conversations/messages/${conversationInfo.id}/communications/${conversationInfo.getCommunicationId(participant!!)}/typing", payload = payload)
}

fun API.sendOutboundMessageFromAgentToUser(
    conversationInfo: Conversation,
    message: String
) {
    val agentParticipant = conversationInfo.getParticipantFromPurpose("agent")
    val communicationId = conversationInfo.getCommunicationId(agentParticipant!!)
    val payload = "{\"textBody\": \"${message}\"}".toByteArray()
    publicApiCall("POST", "/api/v2/conversations/messages/${conversationInfo.id}/communications/$communicationId/messages", payload)
}

fun API.answerNewConversation(): Conversation? {
    changePresence("On Queue", testConfig.agentUserId)
    val conversation = waitForConversation()
    if (conversation == null) {
        println("Failed to receive a new conversation to answer.")
        return null
    }
    sendConnectOrDisconnect(conversation, connecting = true, wrapup = false)
    changePresence("Available", testConfig.agentUserId)
    return getConversationInfo(conversation.id)
}

fun API.sendConnectOrDisconnect(
    conversationInfo: Conversation,
    connecting: Boolean = false,
    wrapup: Boolean = true
) {
    var statusPayload = ""
    println("Sending $connecting request to: ${conversationInfo.id} from a conversation.")
    val agentParticipant = conversationInfo.getParticipantFromPurpose("agent")
    var wrapupCodePayload: JsonElement? = null
    if (!connecting) {
        statusPayload = "{\"state\": \"DISCONNECTED\"}"
        wrapupCodePayload = getDefaultWrapupCodeId(conversationInfo.id, agentParticipant!!.id)
    } else {
        statusPayload = "{\"state\": \"CONNECTED\"}"
    }
    sendStatusToParticipant(conversationInfo.id, agentParticipant!!.id, statusPayload)
    if (connecting) {
        waitForParticipantToConnectOrDisconnect(conversationInfo.id, connectionState = "connected")
    } else if (wrapup) {
        sleep(2000)
        print("Sending wrapup code.")
        sendWrapupCode(conversationInfo, wrapupCodePayload)
    }
}

fun API.getDefaultWrapupCodeId(
    conversationId: String,
    participantId: String
): JsonElement? {
    val wrapupCodeId = publicApiCall("GET", "/api/v2/conversations/$conversationId/participants/$participantId/wrapupcodes")
        ?.jsonArray?.firstOrNull()?.jsonObject?.get("id")
    return wrapupCodeId
}

fun API.waitForParticipantToConnectOrDisconnect(
    conversationId: String,
    connectionState: String = "disconnected"
) {
    val deadline = System.currentTimeMillis() + 60_000L
    while (System.currentTimeMillis() < deadline) {
        try {
            val listOfMessages = getConversationInfo(conversationId).getParticipantFromPurpose("agent")?.messages?.toList()
            if (listOfMessages?.get(0)?.state == connectionState) return
        } catch (_: Exception) {
            // ignore and retry
        }
        sleep(1000)
    }
    throw AssertionError("Participant did not reach state '$connectionState' within 60 seconds")
}

fun API.sendStatusToParticipant(
    conversationId: String,
    participantId: String,
    statusPayload: String
) {
    publicApiCall("PATCH", "/api/v2/conversations/messages/$conversationId/participants/$participantId", statusPayload.toByteArray())
}

fun API.sendWrapupCode(
    conversationInfo: Conversation,
    wrapupPayload: JsonElement?
) {
    val agentParticipant = conversationInfo.getParticipantFromPurpose("agent")
    val communicationId = conversationInfo.getCommunicationId(agentParticipant!!)
    val codeId = wrapupPayload?.jsonPrimitive?.content
    publicApiCall(
        "POST",
        "/api/v2/conversations/messages/${conversationInfo.id}/participants/${agentParticipant.id}/communications/$communicationId/wrapup",
        ("{ \"code\": \"$codeId\", \"notes\": \"\"}").toByteArray()
    )
}

fun API.disconnectAllConversations() {
    val conversationList = getAllConversations()
    conversationList.forEach { conversation ->
        Log.i(TAG, "disconnecting conversationId: ${conversation.id}")
        sendConnectOrDisconnect(conversation)
    }
}

fun API.checkForConversationMessages(conversationId: String) {
    val listOfMessages = getConversationInfo(conversationId).getParticipantFromPurpose("agent")?.messages?.toList()
    if (listOfMessages != null) AssertionError("Conversation still has messages associated with it but should not")
}

fun API.attachImage(conversationInfo: Conversation) {
    val agentParticipant = conversationInfo.getParticipantFromPurpose("agent")
    val communicationId = conversationInfo.getCommunicationId(agentParticipant!!)
    val mediaResult = publicApiCall(
        "POST",
        "/api/v2/conversations/messages/${conversationInfo.id}/communications/$communicationId/messages/media"
    )
    val result: Media = parseJsonToClass(mediaResult)
}
